src/main/scala/io/iohk/ethereum/mining/BlockGenerator.scala

```diff
@@ -37,7 +37,10 @@ class BlockGenerator(blockchainStorages: BlockchainStorages, blockchainConfig: B
         val blockTimestamp = Instant.now.getEpochSecond
         val header: BlockHeader = prepareHeader(blockNumber, ommers, beneficiary, parent, blockTimestamp)
 
-        val transactionsForBlock = transactions.toSet[SignedTransaction].toList.sortBy(_.tx.nonce)
+        val transactionsForBlock = transactions
+          .filter(_.tx.gasLimit < header.gasLimit)
+          //if we have 2 transactions from same address we want first one with lower nonce
+          .sortBy(_.tx.nonce)
```

Comment:
ordering by nonce is to prevent tx failure when we have 2 transactions from same account
this is not optimisation for mining, we will do that in separate PR